bip-xxxx.mediawiki

```diff
@@ -0,0 +1,335 @@
+
+===Copyright===
+
+This BIP is licensed under the 2-clause BSD license.
+
+===Motivation===
+
+When two parties (later referred to as sender and receiver) want to transact,
+most of the time, the sender creates a transaction spending their own Unspent Transaction Outputs (UTXOs), signs
+it and broadcasts it on the network.
+
+This simple model gave birth to several heuristics impacting the privacy of the parties and of the network as a whole.
+
+* Common input ownership heuristic: In most transactions, all the inputs belong to the same party.
+* Change identification from scriptPubKey type: If all inputs are spending UTXOs of a certain scriptPubKey type, then the change output is likely to have the same scriptPubKey type, too.
+* Change identification from round amount: If an output in the transaction has a round amount, it is likely an output belonging to the receiver.
+
+We will designate these three heuristics as <code>common-input</code>, <code>change-scriptpubkey</code>, <code>change-round-amount</code>.
+
+The problems we aim to solve are:
+* For the receiver, there is a missed opportunity to consolidate their own UTXOs or making payment in the sender's transaction.
+* For the sender, there are privacy leaks regarding their wallet that happen when someone applies the heuristics detailed above to their transaction.
+
+Our proposal gives an opportunity for the receiver to consolidate their UTXOs while also batching their own payments, without creating a new transaction. (Saving fees in the process)
+For the sender, it allows them to invalidate the three heuristics above. With the receiver's involvement, the heuristics can even be poisoned. (ie, using the heuristics to intentionally mislead blockchain analysis)
+
+Note that the existence of this proposal is also improving the privacy of parties who are not using it by making the three heuristics unreliable to the network as a whole.
+
+=== Relation to BIP79 (Bustapay) ===
+
+Another implementation proposal has been written: [[https://github.com/bitcoin/bips/blob/master/bip-0079.mediawiki|BIP79 Bustapay]].
+
+We decided to deviate from it for several reasons:
+* It was not using PSBT, so if the receiver wanted to bump the fee, they would need the full UTXO set.
+* The receiver was responsible to pay the additional fee, not the sender.
+* It was requiring at least one input to be contributed by the receiver.
+* Inability to change the payment output to match scriptPubKey type.
+* Lack of basic versioning negotiation if the protocol evolves.
+* No standardization of error condition for proper feedback to the sender.
+
+Other than that, our proposal is very similar.
+
+==Specification==
+
+===Protocol===
+
+In a payjoin payment, the following steps happen:
+
+* The receiver of the payment, presents a [[bip-021.mediawiki|BIP 21 URI]] to the sender with a parameter <code>pj</code> describing an https (or http if it is a Tor hidden service) link to the payjoin endpoint.
```

Comment:
Just some "out there" thoughts. What if the `pj` key is not strictly defined as an "http(s)" endpoint but as "an instruction of establishing a communication with the receiver's endpoint". This would enable receiver scenarios of devices not having a reachable tor or http endpoint such as using QR codes, Bluetooth, NFC, etc


QR Code Payjoin

* Receiver generates BIP21 payment request with a payjoin flag: bitcoin:abc?amount=1&pj=qrcode. Sender shows a QR code with bip21
* Sender scans BIP21 and recognizes qrcode pj flag. Sender generates an original PSBT and encodes it in qr code
* Receiver scans QR code, generates ne PJ PSBT, shows it in QR
* Sender scans, signs, broadcasts

Bluetooth Payjoin

* Receiver generates BIP21 payment request with a payjoin flag: `bitcoin:abc?amount=1&pj=bluetooth:devicefingerprint`. Sender shows a QR code with bip2 ( or BIP21 is transmitted through bluetoooth on an already establish connection)
* Sender scans BIP21 and recognizes bluetoothpj flag. Sender scans bluetooth devices for identifier, connects and sends original psbt
* Receiver generates new PJ PSBT, and sends it back to connected device
* Sender signs, broadcasts

NFC Payjoin

* Receiver generates BIP21 payment request with a payjoin flag: `bitcoin:abc?amount=1&pj=nfc`. Sender shows a QR code with bip21 (or sender taps device with receiver's nfc device)
* Sender scans BIP21 and recognizes NFCpj flag. Sender scans bluetooth devices for identifier, connects and sends original psbt
* Receiver generates new PJ PSBT, and sends it back to connected device with NFC tap
* Sender signs, broadcasts

Reply:
> Just some "out there" thoughts. What if the pj key is not strictly defined as an "http(s)" endpoint but as "an instruction of establishing a communication with the receiver's endpoint". This would enable receiver scenarios of devices not having a reachable tor or http endpoint such as using QR codes, Bluetooth, NFC, etc

But I thought the point of line https://github.com/bitcoin/bips/pull/923/files#diff-bab55ce4db24c444e852baf3d7ddfefaR99 was to state that https/onion are *examples* of urls, and another scheme could be put there? (i.e. I thought the doc already encompassed this eventuality like bluetooth etc)

Reply:
It does. But I think in the context of the sentence, this was not clear.

I relaxed the authenticated channel because I think it is safe as long as no address substitution is allowed... what do you think?

Reply:
Yes, I saw. It makes sense but I probably wouldn't add that to the doc. If people have a way to be secure without following the recommendation, they can, but that would be like a custom implementation, in my view it's better not to make it "official" that transmitting over non-encrypted-and-authenticated channels is "supported" somehow as part of the publically-consensus protocol for payjoin.

Reply:
fair point, will not mention.